atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TopListFilteredCounter.java

```diff
@@ -0,0 +1,123 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import com.codahale.metrics.Counter;
+import com.google.common.collect.Sets;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.time.Duration;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import one.util.streamex.EntryStream;
+
+final class TopListFilteredCounter<T> {
+    private static final SafeLogger log = SafeLoggerFactory.get(TopListFilteredCounter.class);
+
+    private final int maxSize;
+    private final TaggedMetricRegistry registry;
+    private final Function<T, MetricName> tagToMetricName;
+
+    private final Map<T, Counter> counters = new ConcurrentHashMap<>();
+    private final Set<T> reportedTags = ConcurrentHashMap.newKeySet();
+
+    private TopListFilteredCounter(
+            int maxSize, TaggedMetricRegistry registry, Function<T, MetricName> tagToMetricName) {
+        this.maxSize = maxSize;
+        this.registry = registry;
+        this.tagToMetricName = tagToMetricName;
+    }
+
+    static <T> TopListFilteredCounter<T> create(
+            int maxSize,
+            Duration initialDelay,
+            Duration resetInterval,
+            Function<T, MetricName> tagToMetricName,
+            TaggedMetricRegistry registry,
+            ScheduledExecutorService executor) {
+        TopListFilteredCounter<T> counter = new TopListFilteredCounter<>(maxSize, registry, tagToMetricName);
+        executor.scheduleAtFixedRate(
+                counter::resilientUpdateMetricsReporting,
+                initialDelay.toNanos(),
+                resetInterval.toNanos(),
+                TimeUnit.NANOSECONDS);
+        return counter;
+    }
+
+    void inc(T tag, long count) {
+        Counter counter = counters.computeIfAbsent(tag, unused -> new Counter());
+        counter.inc(count);
+    }
+
+    private void resilientUpdateMetricsReporting() {
+        try {
+            updateMetricsReporting();
+        } catch (RuntimeException e) {
+            log.error("Failed to update top listed counter metric", e);
+        }
+    }
+
+    private void updateMetricsReporting() {
+        Set<T> nextBatchOfTagsToReport = computeTopListTags();
+
+        deregisterTagsForMetricsProduction(Sets.difference(reportedTags, nextBatchOfTagsToReport));
+        registerTagsForMetricsProduction(Sets.difference(nextBatchOfTagsToReport, reportedTags));
+
+        reportedTags.clear();
```

Comment:
Irrelevant for this PR given it's single threaded and not a hot path, but a thing to watch out for in general - clearing a set requires removing every element (unsurprising), which on a concurrent set requires synchronizing (eventually) on each internal bucket. I often find, for non-realtime low memory constrained things (esp with the new GC) to just create a fresh set and let GC handle the old one in the background